File: app/src/main/java/com/supercilex/robotscouter/ui/teamlist/UpdateDialog.java
package com.supercilex.robotscouter.ui.teamlist;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;

import com.supercilex.robotscouter.BuildConfig;
import com.supercilex.robotscouter.R;

public class UpdateDialog extends DialogFragment implements AlertDialog.OnClickListener {
    private static final String TAG = "UpdateDialog";
    private static final Uri STORE_LISTING_URI =
            Uri.parse("market://details?id=" + BuildConfig.APPLICATION_ID);

    public static void show(FragmentManager manager) {
        if (manager.findFragmentByTag(TAG) == null) new UpdateDialog().show(manager, TAG);
    }

    public static void showStoreListing(Activity activity) {
        activity.startActivity(new Intent(Intent.ACTION_VIEW).setData(STORE_LISTING_URI));
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new AlertDialog.Builder(getContext())
                .setTitle(R.string.update_required_title)
                .setMessage(R.string.update_required_message)
                .setPositiveButton(R.string.update, this)
                .setOnCancelListener(this)
                .setCancelable(false)
                .create();
    }

    @Override
    public void onClick(DialogInterface dialog, int which) {
        showStoreListing(getActivity());
    }

    @Override
    public void onCancel(DialogInterface dialog) {
        getActivity().finish();
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/teamlist/UpdateDialog.kt
package com.supercilex.robotscouter.ui.teamlist

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.FragmentManager
import android.support.v7.app.AlertDialog
import android.text.Html
import androidx.core.net.toUri
import androidx.core.text.parseAsHtml
import com.supercilex.robotscouter.BuildConfig
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.util.ui.DialogFragmentBase
import com.supercilex.robotscouter.util.updateRequiredMessage

class UpdateDialog : DialogFragmentBase(), DialogInterface.OnClickListener {
    override fun onCreateDialog(savedInstanceState: Bundle?) = AlertDialog.Builder(requireContext())
            .setTitle(R.string.update_required_title)
            .setMessage(updateRequiredMessage.parseAsHtml(Html.FROM_HTML_MODE_COMPACT))
            .setPositiveButton(R.string.update_title, this)
            .setOnCancelListener(this)
            .setCancelable(false)
            .create()

    override fun onClick(dialog: DialogInterface, which: Int) = showStoreListing(requireActivity())

    override fun onCancel(dialog: DialogInterface?) = requireActivity().finish()

    companion object {
        private const val TAG = "UpdateDialog"

        private val STORE_LISTING_URI = "market://details?id=${BuildConfig.APPLICATION_ID}".toUri()
        private val LATEST_APK_URI =
                "https://github.com/SUPERCILEX/app-version-history/blob/master/Robot-Scouter/app-release.apk".toUri()

        fun show(manager: FragmentManager) {
            if (manager.findFragmentByTag(TAG) == null) {
                manager.beginTransaction().add(UpdateDialog(), TAG).commitAllowingStateLoss()
            }
        }

        fun showStoreListing(activity: Activity) = try {
            activity.startActivity(Intent(Intent.ACTION_VIEW).setData(STORE_LISTING_URI))
        } catch (e: ActivityNotFoundException) {
            activity.startActivity(Intent(Intent.ACTION_VIEW).setData(LATEST_APK_URI))
        }
    }
}
